In [3]:
# ============================================================
# ATIVIDADE 3 - Heurística Gulosa x Solução Ótima
# Objetivo: comparar uma heurística simples com a solução ótima
# ============================================================

import itertools
import time
import random
import numpy as np


# ----------------------------------------------------------
# 1. FUNÇÃO PARA RESOLVER A MOCHILA POR FORÇA BRUTA
# ----------------------------------------------------------

def mochila_forca_bruta(pesos, valores, capacidade):
    """
    Testa todas as combinações possíveis e retorna o melhor valor.
    """
    n = len(pesos)

    melhor_valor = 0
    melhor_combinacao = None

    for combinacao in itertools.product([0, 1], repeat=n):

        peso_total = 0
        valor_total = 0

        for i in range(n):
            if combinacao[i] == 1:
                peso_total += pesos[i]
                valor_total += valores[i]

        if peso_total <= capacidade and valor_total > melhor_valor:
            melhor_valor = valor_total
            melhor_combinacao = combinacao

    return melhor_valor, melhor_combinacao


# ----------------------------------------------------------
# 2. HEURÍSTICA GULOSA
# ----------------------------------------------------------

def mochila_gulosa(pesos, valores, capacidade):
    """
    Resolve a mochila utilizando uma heurística gulosa.

    Os itens são ordenados pela razão valor/peso.
    """

    n = len(pesos)

    # Calcula valor/peso de cada item
    itens = []

    for i in range(n):
        if pesos[i] > 0:
            razao = valores[i] / pesos[i]
        else:
            razao = 0

        itens.append((razao, i))

    # Ordena da maior razão para a menor
    itens.sort(reverse=True)

    peso_total = 0
    valor_total = 0
    combinacao = [0] * n

    # Escolhe os itens enquanto houver espaço
    for razao, i in itens:

        if peso_total + pesos[i] <= capacidade:
            combinacao[i] = 1
            peso_total += pesos[i]
            valor_total += valores[i]

    return valor_total, tuple(combinacao)


# ----------------------------------------------------------
# 3. FUNÇÃO PARA CALCULAR O GAP
# ----------------------------------------------------------

def calcular_gap(valor_heuristica, valor_otimo):
    """
    Calcula o gap percentual entre a solução heurística
    e a solução ótima.
    """

    if valor_otimo == 0:
        return 0

    return ((valor_otimo - valor_heuristica) / valor_otimo) * 100


# ----------------------------------------------------------
# 4. GERANDO INSTÂNCIAS ALEATÓRIAS
# ----------------------------------------------------------

np.random.seed(42)

NUM_INSTANCIAS = 20
NUM_ITENS = 12
CAPACIDADE = 30

gaps = []

print("Rodando 20 instancias...")
print("=" * 75)


# ----------------------------------------------------------
# 5. EXECUTANDO AS 20 INSTÂNCIAS
# ----------------------------------------------------------

for k in range(NUM_INSTANCIAS):

    # Gera pesos aleatórios
    pesos = np.random.randint(1, 11, size=NUM_ITENS)

    # Gera valores aleatórios
    valores = np.random.randint(1, 21, size=NUM_ITENS)

    # ------------------------------------------------------
    # Solução ótima
    # ------------------------------------------------------

    inicio = time.time()

    otimo, combinacao_otima = mochila_forca_bruta(
        pesos,
        valores,
        CAPACIDADE
    )

    tempo_otimo = time.time() - inicio

    # ------------------------------------------------------
    # Solução heurística gulosa
    # ------------------------------------------------------

    inicio = time.time()

    heur, combinacao_heuristica = mochila_gulosa(
        pesos,
        valores,
        CAPACIDADE
    )

    tempo_heur = time.time() - inicio

    # ------------------------------------------------------
    # Calculando o GAP
    # ------------------------------------------------------

    gap = calcular_gap(heur, otimo)

    gaps.append(gap)

    # ------------------------------------------------------
    # Mostrando resultado da instância
    # ------------------------------------------------------

    print(
        f"Instancia {k+1:2d} | "
        f"Otimo: {otimo:4d} | "
        f"Gulosa: {heur:4d} | "
        f"Gap: {gap:5.1f}%"
    )


# ----------------------------------------------------------
# 6. ESTATÍSTICAS FINAIS
# ----------------------------------------------------------

print("=" * 75)

print(f"Gap medio: {np.mean(gaps):.2f}%")
print(f"Maior gap: {np.max(gaps):.2f}%")
print(f"Menor gap: {np.min(gaps):.2f}%")

print("=" * 75)


# O que entregar / discutir:
# 19.	Codigo completo (com a funcao calcular_gap implementada e o loop funcionando).
# 20.	Valor do gap medio obtido.
# 21.	Resposta: “A heuristica gulosa e boa o suficiente para este problema? Em quais situacoes voce usaria ela e em quais preferiria gastar mais tempo para achar o otimo?”

Rodando 20 instancias...
Instancia  1 | Otimo:   72 | Gulosa:   70 | Gap:   2.8%
Instancia  2 | Otimo:   87 | Gulosa:   87 | Gap:   0.0%
Instancia  3 | Otimo:  104 | Gulosa:  104 | Gap:   0.0%
Instancia  4 | Otimo:   66 | Gulosa:   66 | Gap:   0.0%
Instancia  5 | Otimo:   65 | Gulosa:   65 | Gap:   0.0%
Instancia  6 | Otimo:   83 | Gulosa:   80 | Gap:   3.6%
Instancia  7 | Otimo:   83 | Gulosa:   82 | Gap:   1.2%
Instancia  8 | Otimo:   79 | Gulosa:   79 | Gap:   0.0%
Instancia  9 | Otimo:  117 | Gulosa:  117 | Gap:   0.0%
Instancia 10 | Otimo:   90 | Gulosa:   85 | Gap:   5.6%
Instancia 11 | Otimo:  102 | Gulosa:  102 | Gap:   0.0%
Instancia 12 | Otimo:   99 | Gulosa:   99 | Gap:   0.0%
Instancia 13 | Otimo:  100 | Gulosa:  100 | Gap:   0.0%
Instancia 14 | Otimo:   91 | Gulosa:   87 | Gap:   4.4%
Instancia 15 | Otimo:   81 | Gulosa:   81 | Gap:   0.0%
Instancia 16 | Otimo:   84 | Gulosa:   84 | Gap:   0.0%
Instancia 17 | Otimo:   79 | Gulosa:   77 | Gap:   2.5%
Instancia 18 | Otimo:  

## Tabela de Resultados — LAB-03

| Instância | Solução Ótima | Solução Gulosa | Gap (%) |
|:---------:|--------------:|---------------:|--------:|
| 1  | 72  | 70  | 2,8 |
| 2  | 87  | 87  | 0,0 |
| 3  | 104 | 104 | 0,0 |
| 4  | 66  | 66  | 0,0 |
| 5  | 65  | 65  | 0,0 |
| 6  | 83  | 80  | 3,6 |
| 7  | 83  | 82  | 1,2 |
| 8  | 79  | 79  | 0,0 |
| 9  | 117 | 117 | 0,0 |
| 10 | 90  | 85  | 5,6 |
| 11 | 102 | 102 | 0,0 |
| 12 | 99  | 99  | 0,0 |
| 13 | 100 | 100 | 0,0 |
| 14 | 91  | 87  | 4,4 |
| 15 | 81  | 81  | 0,0 |
| 16 | 84  | 84  | 0,0 |
| 17 | 79  | 77  | 2,5 |
| 18 | 101 | 101 | 0,0 |
| 19 | 94  | 92  | 2,1 |
| 20 | 115 | 113 | 1,7 |
| **Média** | — | — | **1,20** |
| **Maior** | — | — | **5,56** |
| **Menor** | — | — | **0,00** |


A heurística gulosa apresentou um bom desempenho nas 20 instâncias analisadas. Em várias delas, a solução gulosa encontrou exatamente o mesmo valor da solução ótima, resultando em gap de 0%. Nas instâncias em que houve diferença, o gap permaneceu relativamente pequeno.

O gap médio foi de 1,20%, enquanto o maior gap observado foi de 5,56%. Isso mostra que, para essas instâncias, a heurística conseguiu encontrar soluções próximas da ótima sem precisar testar todas as combinações.

Entretanto, a heurística gulosa não garante encontrar a solução ótima em todos os casos. Ela toma decisões com base no melhor valor/peso disponível no momento, podendo escolher itens que impedem uma combinação melhor posteriormente. A força bruta, por outro lado, testa todas as possibilidades e consegue garantir a solução ótima, mas apresenta um custo computacional muito maior conforme o número de itens aumenta.